### PR TITLE
Article: Using Ktor in Kotlin Multiplatform

### PR DESCRIPTION
**Pull request by mohsenr**

### CODE REVIEW

**Review by simonseyer — COMMENTED** [review]
It's a nice article that is easy to read and follow. One thing I would suggest is to add a summary at the end to wrap up the general idea. The end was a bit abrupt.

File: drafts/ktor-in-kmp.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+* The choice of engine and how it is configured will depend on the platform.
+* Internal builds may talk to service at a different address and may have different credentials
+* In internal builds, or during testing, we may want to use an in-app mock instead of a live server.
+* Logging behaviour may depend on the type of build (e.g. would be disabled in production).
+
+We like to decouple how we set these environment-specific configuration from how the app code itself configures its `HttpClient`. In fact, the environment configuration may be in a different module from the app code as we don’t want our internal environment credentials or mocks to be shipped with the app in production.
+
+## Interfacing our problem
+
+Due to the way `HttpClient` is configured, we can’t inject a pre-built instance into the app as it wouldn’t be able to further configure it anymore. We also don’t want to manually inject configuration data objects into the app, as ideally it shouldn’t even know *what* should it configure.
```
Comment (simonseyer) [inline drafts/ktor-in-kmp.md:60]:
```suggestion
Due to the way `HttpClient` is configured, we can’t inject a pre-built instance into the app as it wouldn’t be able to further configure it anymore. We also don’t want to manually inject configuration data objects into the app, as ideally it shouldn’t even know *what* it should configure.
```

File: drafts/ktor-in-kmp.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+For smaller projects, this is probably the right trade-off. For bigger projects or teams, you may find it’s better to strengthen the boundaries by creating custom type-safe API.
+
+## Mocking can be expanded
+
+Implementation of `MockHttpClientProvider` (as written here) can easily get out of hand when you start mocking many APIs, and maybe have different possible responses for each of those APIs, so consider improvements that suit your needs.
+
+# Summary
+
+In this article we explore a method of systematically improving separation of concerns and in turn maintainability of networking infrastructure in a KMP app.
```
Comment (simonseyer) [inline drafts/ktor-in-kmp.md:243]:
```suggestion
In this article we explored a method of systematically improving separation of concerns and in turn maintainability of networking infrastructure in a KMP app.
```

**Review by simonseyer — APPROVED** [review]
Just two nitpicks (didn't check the code in detail)

File: drafts/ktor-in-kmp.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+}
+```
+
+For the “default” use case, we may have an implementation of that interface like so:
+
+```kotlin
+class DefaultHttpClientProvider(private val host: String,
+                                private val path: String? = null,
+                                private val scheme: String = "https") : HttpClientProvider {
+    override fun httpClient(commonInit: HttpClientConfig<*>.() -> Unit): HttpClient {
```
[the remark this answers is not in the document]
Comment (mohsenr) [reply drafts/ktor-in-kmp.md:79]: Thanks, fixed it… I think :)

File: drafts/ktor-in-kmp.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            }
+
+            commonInit(this)
+        }
+    }
+}
+
+internal expect fun newPlatformHttpClient(config: HttpClientConfig<*>.() -> Unit = {}): HttpClient
+```
+
```
[the remark this answers is not in the document]
Comment (mohsenr) [reply drafts/ktor-in-kmp.md:100]: Good suggestion. Done.

**Review by aroyarexs — APPROVED** [review]

**Pull request merged**